### PR TITLE
[std, framework] Adds "buffer" module to std

### PR DESCRIPTION
**Pull request by damirka**
## Description 

Adds `std::buffer` module for aggregating any data.

## Test Plan 

TBD

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes

- adds the `std::buffer` module for aggregating data (intended use within PTB)

### CODE REVIEW

File: crates/sui-framework/packages/move-stdlib/sources/buffer.move [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+            contents: vector[]
+        }
+    }
+
+    /// Append data to the buffer.
+    public fun append<T>(self: &mut Buffer<T>, data: vector<T>) {
+        vector::append(&mut self.contents, data)
+    }
+
+    /// Unwrap the buffer and return the contents.
```
Comment (tzakian) [inline crates/sui-framework/packages/move-stdlib/sources/buffer.move:41]: Thoughts on also adding a way to access the size of the buffer? Or do you want to keep anything about the contents truly opaque during construction?

**Review by tzakian — COMMENTED** [review]
This looks fine to me, although I'm not necessarily sold on how useful this will be in practice vs just doing manual size checks? Is there a particular place or example that you could point me to where I could see where this might be useful?

**Comment by tnowacki** [comment]
I have similar concerns as @tzakian about the general usefulness. 

I think also that long term this feels like something you might just want a macro for with `vector`

Something like 
```
macro fun exact_size<T>(n: u64, $f: |&mut vector<T>|) {
  let v = vector[];
  $f(&mut v);
  assert!(v.length() == n, 42);
}
```

**Comment by amnn** [comment]
@tnowacki, I believe the background here is that `Buffer` is used to accumulate values on chain when the end result is bigger than could fit in a single transaction payload (like sending parts up to the ISS in separate rockets and assembling it in space), so a macro wouldn't work.

In terms of general usefulness, I'm not sure, I would tend to rely on @damirka's anecdotal evidence for that, given that I think he's used this in anger multiple times already, but I did have similar reservations to other folks about whether it is so universally useful to be included in `stdlib` or `sui`.

Given it's free-standing, I would much prefer for us to work on a policy for us to share utility libraries freely on-chain, but under their own Package IDs, and come up with a solution to address the trust aspect of depending on a random on-chain package.

**Comment by damirka** [comment]
Needs more discussion, closing for now.

**Pull request closed without merging**